Fix Input resolution for inputs only present in db

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1670

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 1502136da13c0f9e6ba47b74c12c31912099da62
  :round_pushpin: **add regression test illustrating the issue**
  Input that should be resolved aren't. In order to resolve inputs, we are actually wrongly using
the transactions available from the history that was fetched from the database, instead of using
the entire database! So, giving different query parameters restricting the set of outputs has an
influence on something it shouldn't.

- 8ec012359e5faaeeec9f73f255184575b6b5d0c1
  :round_pushpin: **resolve transaction inputs from db when fetching the history**
  Before, this was done in the application layer, from the available history ('assemble' in 'Cardano.Wallet'). This requires exactly two extra lookups. One in the TxOut table, and one in the Checkpoint table so the performance impact should be quite small. It simplifies the application logic quite a bit and doesn't much impact the database logic, so it's almost a win-win. Plus, it fixes the issue where resolvable inputs would appear as non resolved

- 64bf4db742dc029ea3a18b1c4e957e22f612ab47
  :round_pushpin: **fix DB model to also return full 'TransactionInfo' when fetching the tx history**
  
- 21d75d20ae9b11707d16a4115159ef617e6bcdb0
  :round_pushpin: **Fix db benchmark build**
  Benchmarks may produce higher timings after this change, but it's OK
since the answers are more correct now, and the work has just shifted
from the wallet layer into the db layer.

- fffe476a58673fcb23a3c43136d8eded2d46f999
  :round_pushpin: **re-add lost comment about input resolution for incoming/outgoing transactions**
  
- 67adc07a71a3e8c3e5430654b0845173e4d189e2
  :round_pushpin: **rename 'txs' -> 'pending' to better match context**


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
add regression test illustrating the issue

Input that should be resolved aren't. In order to resolve inputs, we are actually wrongly using
the transactions available from the history that was fetched from the database, instead of using
the entire database! So, giving different query parameters restricting the set of outputs has an
influence on something it shouldn't.

KtorZ committed:
resolve transaction inputs from db when fetching the history

Before, this was done in the application layer, from the available history ('assemble' in 'Cardano.Wallet'). This requires exactly two extra lookups. One in the TxOut table, and one in the Checkpoint table so the performance impact should be quite small. It simplifies the application logic quite a bit and doesn't much impact the database logic, so it's almost a win-win. Plus, it fixes the issue where resolvable inputs would appear as non resolved

KtorZ committed:
fix DB model to also return full 'TransactionInfo' when fetching the tx history

KtorZ commented on lib/byron/test/integration/Test/Integration/Byron/Scenario/API/Transactions.hs:
```diff
@@ -690,6 +696,71 @@ scenario_TRANS_UTXO_01 fixtureSource fixtureTarget = it title $ \ctx -> do
+        )
+    => (Context t -> [Natural] -> IO ApiByronWallet)
+    -> SpecWith (Context t)
+scenario_TRANS_REG_1670 fixture = it title $ \ctx -> do
```
This fails before the fix.

rvl replied:
Since the tx input resolving is moved into the DBLayer, this test could probably be more easily done at the unit level.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1493,45 +1493,6 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
-    -- to assemble coin selection information for outgoing payments.
-    -- To reliably provide this information, it should be looked up when
-    -- applying blocks, but that is future work (issue #573).
-    assemble
```
no longer needed, the db now returns `[TransactionInfo]` directly.

rvl replied:
Is it still noted somewhere that coin selection information can't (yet) be provided for outgoing payments?

KtorZ replied:
I've rewritten the comment and linked to the github issue in the relevant place in the db layer :+1: , good point

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1016,15 +1033,35 @@ selectUTxO cp = fmap entityVal <$>
-        outs <- fmap entityVal <$> selectList [TxOutputTxId <-. txids]
+
+        resolvedInputs <- toOutputMap . fmap entityVal <$> selectList
+            [TxOutputTxId <-. (txInputSourceTxId <$> inputs)]
```
If only persistent could support SQL joins (without raw sql).

KtorZ replied:
If only yeah...

rvl commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -599,7 +599,7 @@ readWallet ctx wid = db & \DBLayer{..} -> mapExceptT atomically $ do
     meta <- withNoSuchWallet wid $ readWalletMeta pk
     txs <- lift $ readTxHistory pk Descending wholeRange (Just Pending)
-    pure (cp, meta, Set.fromList (fst <$> txs))
+    pure (cp, meta, Set.fromList (fromTransactionInfo <$> txs))
```
Is it possible to rename `txs` here to `pending`?

rvl reviewed (approved):
LGTM

rvl committed:
Fix db benchmark build

Benchmarks may produce higher timings after this change, but it's OK
since the answers are more correct now, and the work has just shifted
from the wallet layer into the db layer.

KtorZ committed:
re-add lost comment about input resolution for incoming/outgoing transactions

KtorZ committed:
rename 'txs' -> 'pending' to better match context
